src/data/nat/nth.lean

```diff
@@ -0,0 +1,406 @@
+/-
+Copyright (c) 2021 Vladimir Goryachev. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Vladimir Goryachev, Kyle Miller, Scott Morrison, Eric Rodriguez
+-/
+import data.list.basic
+import data.nat.prime
+import data.nat.count
+import order.order_iso_nat
+import set_theory.fincard
+
+/-!
+# The `n`th Number Satisfying a Predicate
+
+This file defines a function for "what is the `n`th number that satisifies a given predicate `p`",
+and provides lemmas that deal with this function and its connection to `nat.count`.
+
+## Main definitions
+
+* `nth p n`: The `n`-th natural `k` (zero-indexed) such that `p k`. If there is no
+  such natural (that is, `p` is true for at most `n` naturals), then `nth p n = 0`.
+
+## Main results
+
+* `nat.nth_eq_order_iso_of_nat`: An infinite set of natural numbers is order-isomorphic to the
+  natural numbers.
+
+## TODO
+
+There has been some discussion on the subject of whether both of `nth` and
+`nat.subtype.order_iso_of_nat` should exist. See discussion
+[here](https://github.com/leanprover-community/mathlib/pull/9457#pullrequestreview-767221180).
+Future work should address how lemmas that use these should be written.
+
+-/
+
+open finset
+
+namespace nat
+variable (p : ℕ → Prop)
+
+/-- Find the `n`-th natural number satisfying `p` (indexed from `0`, so `nth p 0` is the first
+natural number satisfying `p`), or `0` if there is no such number. See also
+`subtype.order_iso_of_nat` for the order isomorphism with ℕ when `p` is infinitely often true. -/
+noncomputable def nth : ℕ → ℕ
+| n := Inf { i : ℕ | p i ∧ ∀ k < n, nth k < i }
+
+instance decidable_pred_mem_set_of [h : decidable_pred p] : decidable_pred (∈ set_of p) := h
```

Comment:
We already have https://leanprover-community.github.io/mathlib_docs/data/set/basic.html#set.decidable_set_of
```suggestion

instance decidable_pred_mem_set_of [h : decidable_pred p] : decidable_pred (∈ set_of p) := h
```

Reply:
Cool, deleted